# **Waze Project**

## Build a Dataframe 
Importing of pandas and numpy other than the actual dataframe we are gonna be working on.

In [6]:
import pandas as pd
import numpy as np

df = pd.read_csv('/Users/edoardochiari/Desktop/Google Projects/CSV/waze_app_dataset.csv')

## Inspect the data 
View and inspect summary informations about the dataframe

In [3]:
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [7]:
df.describe()

ID      sessions        drives  total_sessions  \
count  14999.000000  14999.000000  14999.000000    14999.000000   
mean    7499.000000     80.633776     67.281152      189.964447   
std     4329.982679     80.699065     65.913872      136.405128   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.661156   
50%     7499.000000     56.000000     48.000000      159.568115   
75%    11248.500000    112.000000     93.000000      254.192341   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14999.000000            14999.000000   
mean               1749.837789              121.605974   
std                1008.513876              148.121544   
min                   4.000000                0.000000   
25%                 878.000000                9.000000   
50%                1741.000000               71.000000   
75%                2623.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14999.000000      14999.000000             14999.000000   
mean                29.672512       4039.340921              1860.976012   
std                 45.394651       2502.149334              1446.702288   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2212.600607               835.996260   
50%                  9.000000       3493.858085              1478.249859   
75%                 43.000000       5289.861262              2464.362632   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14999.000000  14999.000000  
mean       15.537102     12.179879  
std         9.004655      7.824036  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

From the output of the `head` function, we can observe that the first 10 rows contain no missing values. However, after running the `info` function, it becomes clear that the dataset does have missing values, specifically 700 in the 'label' column.

Regarding data types, the dataset consists of 13 columns: three of type `float64`, eight of type `int64`, and two of type `object`, with a total of 14,999 rows.

## Understand the data
Given the large number of null values, a good starting point would be to investigate whether there is a noticeable difference between a sample with no null values and one consisting entirely of null values in the 'label' column.

In [8]:
# Isolate rows with null values
null_df = df[df['label'].isnull()]
# Display summary stats of rows with null values
null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [10]:
# Isolate rows without null values
not_null_df = df[~df['label'].isnull()]
# Display summary stats of rows without null values
not_null_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

Comparing the summary statistics of the two groups reveals a very low difference between the two samples due to the means and standard deviations being relatively similar to eachother.

Another important step would be to compare the two populations based on device type, determining whether iPhone or Android users had the majority of the null values.

In [14]:
# Get count of null values by device
null_df['device'].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

In [18]:
# Calculate % of iPhone nulls and Android nulls
null_df['device'].value_counts(normalize=True)

device
iPhone     0.638571
Android    0.361429
Name: proportion, dtype: float64

Out of the 700 rows with null values, 447(64%) were iPhone users and 253(36%) were Android users. However how would this compare to the device ratio in the full dataset?

In [22]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)

device
iPhone     0.644843
Android    0.355157
Name: proportion, dtype: float64

The percentage of the missing values by each device is consistent with their representation in the data overall.

What about examining the count and percentages of users who chured vs. those who retained?

In [26]:
# Calculate counts of churned vs. retained
print(df['label'].value_counts())
print()
print(df['label'].value_counts(normalize=True))

label
retained    11763
churned      2536
Name: count, dtype: int64

label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64


The dataset contains 11763(82%) retained and 2536(18%) churned users.

To better understand the differences between churned and retained users, the median of each variable should be compared, rather than the mean. The median provides a more accurate representation of the typical user by minimizing the impact of outliers, which could skew the results if we used the mean. This approach ensures that extreme values do not distort the comparison, offering a clearer insight into the central tendencies of the churned versus retained users.

In [30]:
# Calculate median values of all columns for churned and retained users
df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

Churned users, on average, completed about three more drives in the last month than those who were retained, but retained users accessed the app on more than twice as many days during that same period. 

When looking at the median values, churned users drove approximately 200 more kilometers and spent 2.5 additional hours on the road compared to retained users. 

This suggests that churned users had more drives condensed into fewer days, with longer and farther trips. These patterns could potentially indicate a distinct user profile, where churned users exhibit higher-intensity app usage over shorter periods, which may contribute to their eventual churn.

To further the study, a good idea could be to clauclate the median kilometers per drive for both retained and churned users:

In [35]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median, and isolate for km per drive
median_km_per_drive = df.groupby('label').median(numeric_only=True)[['km_per_drive']]
median_km_per_drive

km_per_drive
label                 
churned      74.109416
retained     75.014702

The median of the churned user is about one kilometer shorter with respect to the retained one. How many kilometers per driving day was this? What about the number of drives per driving day?

In [43]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median for both columns at once
median_values = df.groupby('label').median(numeric_only=True)[['km_per_driving_day', 'drives_per_driving_day']]

median_values

km_per_driving_day  drives_per_driving_day
label                                               
churned           697.541999                 10.0000
retained          289.549333                  4.0625

The median churned user drove an impressive 698 kilometers on each day they were active last month, which is approximately 240% more than the median retained user’s per-drive-day distance. In addition, the churned users exhibited a similarly disproportionate number of drives per day compared to retained users.

These figures highlight that, regardless of churn status, the users in this dataset are exceptionally heavy drivers. This suggests that the data may not reflect the behavior of typical drivers; instead, it likely includes a significant number of professional or long-haul drivers. In particular, the high driving volume among churned users points to the possibility that this group contains a large proportion of long-haul truckers.

Given the intensity of their driving habits, it would be valuable for Waze to gather more targeted data on these "super-drivers." Their extensive driving may be linked to unmet needs that differ from those of the average driver, such as daily commuters. Understanding why these users drive so much could reveal insights into why the Waze app may not fully meet their unique requirements, potentially offering opportunities for tailored improvements.

In conclusion, an examination on whether there is an imbalance in how many users churned by device type, could provide useful insights.

In [49]:
# Calculate the count of Android and iPhone users for each label
device_counts = df.groupby(['label', 'device']).size().unstack()

# Calculate the percentage of Android and iPhone users for each label
device_percentages = df.groupby('label')['device'].value_counts(normalize=True).unstack() * 100

# Merge the counts and percentages into a single table
combined_table = pd.concat([device_counts, device_percentages], axis=1, keys=['Count', 'Percentage'])

# Display the final table
combined_table


Count        Percentage           
device   Android iPhone    Android     iPhone
label                                        
churned      891   1645  35.134069  64.865931
retained    4183   7580  35.560656  64.439344

The ratio of iPhone to Android users remains consistent across both the churned and retained groups, mirroring the ratio found in the overall dataset. This suggests that the distribution of device types does not significantly differ between users who churn and those who remain active, indicating that device preference is not a notable factor in user retention or churn in this case.

## Conclusion
The dataset contained 700 missing values in the "label" column, with no discernible pattern to the missing data. When analyzing the data, the median was preferred over the mean because the median is less affected by outliers, providing a better representation of a typical user in the presence of extreme values.

Further questions arose from the investigation, particularly regarding the driving behavior of churned users. For example, the median churned user drove 698 kilometers per driving day, about 240% more than retained users. It would be important to understand how this data was collected and whether it reflects a non-random sample of users.

Regarding device distribution, about 36% of users were Android users, while 64% were iPhone users. However, there was no significant difference in churn rates between the two groups; the churn rates for iPhone and Android users were within one percentage point of each other, suggesting no correlation between device type and churn.

In terms of distinguishing characteristics, churned users tended to drive farther and for longer durations but did so over fewer days compared to retained users. Additionally, churned users used the app about half as often as retained users during the same period. These patterns point to differences in usage intensity between the two groups.